sys/net/application_layer/nanocoap/sock.c

```diff
@@ -361,6 +361,36 @@ ssize_t nanocoap_sock_post(nanocoap_sock_t *sock, const char *path,
     return _sock_put_post(sock, path, COAP_METHOD_POST, request, len, response, len_max);
 }
 
+static ssize_t _sock_put_post_url(const char *url, unsigned code,
+                                  const void *request, size_t len,
+                                  void *response, size_t len_max)
+{
+    nanocoap_sock_t sock;
+    int res = nanocoap_sock_url_connect(url, &sock);
+    if (res) {
+        return res;
+    }
+
+    res = _sock_put_post(&sock, sock_urlpath(url), code, request, len, response, len_max);
+    nanocoap_sock_close(&sock);
+
+    return res;
+}
+
+ssize_t nanocoap_sock_put_url(const char *url,
+                              const void *request, size_t len,
+                              void *response, size_t len_max)
```

Comment:
If I recall correctly, a response to a PUT request may not hold a payload. If so, the `response` and `len_max` arguments should be dropped.

It's a pity that the response status code is not accessible. But I guess this is consistent with he non-url variants.

Reply:
https://github.com/Tanganelli/CoAPthon/issues/45#issuecomment-266956051 suggests that PUT can have a payload response, AFAIU it's almost equivalent to POST.

Reply:
The RFC says in [section 5.8.3](https://datatracker.ietf.org/doc/html/rfc7252#section-5.8.3):

>     If a resource exists at the request URI, the enclosed representation
>    SHOULD be considered a modified version of that resource, and a 2.04
>    (Changed) Response Code SHOULD be returned.  If no resource exists,
>    then the server MAY create a new resource with that URI, resulting in
>    a 2.01 (Created) Response Code.  If the resource could not be created
>    or modified, then an appropriate error Response Code SHOULD be sent.

And then for "Created" in [section 5.9.1.1](https://datatracker.ietf.org/doc/html/rfc7252#section-5.9.1.1)

>    Like HTTP 201 "Created", but only used in response to POST and PUT
>    requests.  The payload returned with the response, if any, is a
>    representation of the action result.

and for "Changed" in [section 5.9.1.4](https://datatracker.ietf.org/doc/html/rfc7252#section-5.9.1.4)

>    This Response Code is like HTTP 204 "No Content" but only used in
>    response to POST and PUT requests.  The payload returned with the
>    response, if any, is a representation of the action result.

So yes, payloads are allowed in both cases for PUT.

When in doubt, we should rather use the standard for citing instead of issues of some other implementation. Especially, if the RFC is so nicely unambiguous :-).

Reply:
> the RFC is so nicely unambiguous

Heh, that's where I would disagree 

> The payload returned with the response, if any

could mean that payload is optional or that only one of the two options can have payload. Some additional bit of clarification wouldn't have hurt. :smiley:

Reply:
This "if any" would allow all of the wording be in alignment with "responses to PUT cannot have a payload". As always, reading the RFC leaves one more confused than before.

Reply:
I guess at this point it is clear that there will be servers replying with a payload in the response, even if deep hidden in the RFC there would be evidence that this would not be allowed. So either it is allowed (and we have to support it), or it would be pragmatic to keep it anyway to be compatible with real world implementations.

Reply:
> This "if any" would allow all of the wording be in alignment with "responses to PUT cannot have a payload". As always, reading the RFC leaves one more confused than before.

Mh there is no "PUT MUST NOT contain any payload", so I read the "if any" as "if there is a payload"...

Reply:
In HTTP a response to PUT is allowed: https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/PUT

How I wish such documentation would be available for all protocols.

Reply:
I think we agree that PUT can have payload.